### PR TITLE
Add review page for 1995 form

### PR DESCRIPTION
**Pull request by jbalboni**
Related to #4608 

This adds the review page for the new form work. It will need some tweaking as we go (especially with growable tables), but this should work for most of the form. It also sets up Redux, since that was necessary to get the form data onto the review page.

The way the review page works is that it still uses the library to render each form page, but when in review mode it replaces the fields with read-only versions. You can see these in the schemaform/review folder.

I had to change the markup for the read view of each page to be a definition list instead of a table. You can't nest `tbody` elements so it doesn't mesh will with the way the form library recursively generates forms. Caitlin had mentioned wanting to do this in other issue anyway.

Screenshots were taken with random chapters/pages I added in the form config, then removed.

Submit process works (but we don't have an endpoint yet)
screen shot 2017-01-10 at 4 16 18 pm

Collapsed review page
screen shot 2017-01-10 at 3 59 14 pm

Privacy agreement set up done
screen shot 2017-01-10 at 3 59 04 pm

Review mode of expanded chapter
screen shot 2017-01-10 at 3 58 26 pm

Expanded chapter in edit mode
screen shot 2017-01-10 at 3 58 53 pm

### CODE REVIEW

File: src/js/common/schemaform/actions.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+export function setSubmitted(response) {
+  return {
+    type: SET_SUBMITTED,
+    response
+  };
+}
+
+export function submitForm(formConfig, form) {
```
Comment (jbalboni) [inline src/js/common/schemaform/actions.js]: This whole function should be generic, but there's some edu specific stuff in here for now.

File: src/js/common/schemaform/review/ReviewCollapsibleChapter.jsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            <div className="accordion-header clearfix" >
+              <button
+                  className="usa-button-unstyled"
+                  aria-expanded={this.state.open ? 'true' : 'false'}
+                  aria-controls={`collapsible-${this.id}`}
+                  onClick={this.toggleChapter}>
+                {this.props.chapter.title}
+              </button>
+            </div>
+            {this.state.open && pageContent}
```
Comment (akainic) [inline src/js/common/schemaform/review/ReviewCollapsibleChapter.jsx]: Do you need to have this check for `this.state.open` here if you have it above before setting the value of `pageContent`?

**Review by akainic — COMMENTED** [review]
Wow, so much great work! Just a few small comments/questions.

**Comment by jbalboni** [comment]
@akainic All fixed now!

File: src/js/common/components/form-elements/ErrorableCheckbox.jsx [hunk 3]
```diff
@@ -90,7 +90,7 @@ ErrorableCheckbox.propTypes = {
   checked: React.PropTypes.bool,
   errorMessage: React.PropTypes.string,
   name: React.PropTypes.string,
-  label: React.PropTypes.string.isRequired,
+  label: React.PropTypes.object.isRequired,
```
Comment (U-DON) [inline src/js/common/components/form-elements/ErrorableCheckbox.jsx]: Is this an intentional change?

Comment (jbalboni) [inline src/js/common/components/form-elements/ErrorableCheckbox.jsx]: Yes, though only tangentially related to this PR. We sometimes pass rendered components as checkbox labels (so we can use links) and hadn't ever changed this prop type.

Comment (U-DON) [inline src/js/common/components/form-elements/ErrorableCheckbox.jsx]: Then wouldn't labels that are still strings trigger a warning about the type? This should probably be:

```
label: React.PropTypes.oneOfType([
  React.PropTypes.string,
  React.PropTypes.object
]).isRequired,
```

File: src/js/common/schemaform/FormPage.jsx [hunk 4]
```diff
@@ -118,31 +136,50 @@ class FormPage extends React.Component {
...
       </Form>
     );
   }
 }
 
+function mapStateToProps(state, ownProps) {
+  return {
+    form: state.form,
+    route: ownProps.route,
+    reviewMode: ownProps.reviewMode
```
Comment (U-DON) [inline src/js/common/schemaform/FormPage.jsx]: I'm confused why these have to be extracted from `ownProps`; if they were already props being passed in, is this necessary?

File: src/js/common/schemaform/helpers.js [hunk 5]
```diff
@@ -107,13 +124,21 @@ export function formatISOPartialDate({ month, day, year }) {
...
+  return undefined;
+}
 export function parseISODate(dateString) {
   if (dateString) {
     const [year, month, day] = dateString.split('-', 3);
 
     return {
       month: month === 'XX' ? '' : Number(month),
-      day: day === 'XX' ? '' : Number(day),
+      day: day === 'XX' ? '' : Number(day).toString(),
```
Comment (U-DON) [inline src/js/common/schemaform/helpers.js]: Why isn't it also `Number(month).toString()` above this?

File: src/js/common/schemaform/review/ObjectField.jsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+class ObjectField extends React.Component {
+  static defaultProps = {
+    uiSchema: {},
+    errorSchema: {},
+    idSchema: {},
+    registry: getDefaultRegistry(),
+    required: false,
+    disabled: false,
+    readonly: false,
```
Comment (U-DON) [inline src/js/common/schemaform/review/ObjectField.jsx]: It seems some of these props aren't being used at all, like `required`, `disabled`, `readonly`.

File: src/js/common/components/form-elements/ErrorableCheckbox.jsx (hunk 3 above)
Comment (jbalboni) [inline src/js/common/components/form-elements/ErrorableCheckbox.jsx]: Yes, it should be `oneOf`.

File: src/js/common/schemaform/FormPage.jsx (hunk 4 above)
Comment (jbalboni) [inline src/js/common/schemaform/FormPage.jsx]: I was under the impression that react-redux wouldn't pass through props you passed to the connect wrapped component to the inner component. I'm not sure I have tried that directly, so I could be wrong.

File: src/js/common/schemaform/helpers.js (hunk 5 above)
Comment (jbalboni) [inline src/js/common/schemaform/helpers.js]: The way our current date component works (which I copied), it requires the month to be a number and the day to be a string.

File: src/js/common/schemaform/review/ObjectField.jsx (hunk 6 above)
Comment (jbalboni) [inline src/js/common/schemaform/review/ObjectField.jsx]: Yeah, I need to pass those to `SchemaField` or remove them. This component is a copied and modified version of one in the `react-jsonschema-form` library, so we don't necessarily use all of the props that could be passed to it.

Comment (jbalboni) [inline src/js/common/schemaform/review/ObjectField.jsx]: Actually, those are all being used now.

File: src/js/common/schemaform/FormPage.jsx (hunk 4 above)
Comment (jbalboni) [inline src/js/common/schemaform/FormPage.jsx]: I am wrong! I'll remove those.

File: test/common/schemaform/review/ObjectField.unit.spec.jsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+          schema={schema}
+          formContext={{}}
+          idSchema={{ $id: 'root' }}
+          formData={{}}
+          onChange={onChange}
+          onBlur={onBlur}/>
+    );
+
+    expect(tree.everySubTree('.form-review-panel-page-header-row')).not.to.be.empty;
+    expect(tree.subTree('.form-review-panel-page-header').text()).to.be.equal('Blah');
```
Comment (U-DON) [inline test/common/schemaform/review/ObjectField.unit.spec.jsx]: Nitpick: `to.equal`?

File: src/js/common/schemaform/review/ReviewCollapsibleChapter.jsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        </div>
+      );
+    }
+
+    return (
+      <div id={`${this.id}-collapsiblePanel`} className="usa-accordion-bordered form-review-panel">
+        <Element name={`chapter${this.props.chapterKey}ScrollElement`}/>
+        <ul className="usa-unstyled-list">
+          <li>
+            <div className="accordion-header clearfix" >
```
Comment (U-DON) [inline src/js/common/schemaform/review/ReviewCollapsibleChapter.jsx]: Nitpick: extra space.

File: test/common/schemaform/widgets/DateWidget.unit.spec.jsx [hunk 9]
```diff
@@ -63,9 +63,11 @@ describe('Schemaform: DateWidget', () => {
     const instance = tree.getMountedInstance();
     instance.handleChange('year', 2001);
     instance.handleChange('month', '1');
-    instance.handleChange('day', '2');
```
Comment (U-DON) [inline test/common/schemaform/widgets/DateWidget.unit.spec.jsx]: I thought you mentioned for this component, month is expected to be a number and day to be a string? This seems at odds with the above change in this set of tests too.

Comment (jbalboni) [inline test/common/schemaform/widgets/DateWidget.unit.spec.jsx]: Went back and looked at this to refresh my memory. I'm not sure what I was running into before, but it seems to work fine if they're all strings. I'm converting to number first to remove leading zeroes, but the end result seems to be fine in either type.

In this test I'm not actually using these values, so it doesn't break anything, but I'll fix it in both places.

**Review by U-DON — APPROVED** [review]
LGTM.

**Pull request merged**